### PR TITLE
Read config from configMaps referenced by a label in net-attach-def

### PR DESCRIPTION
**Pull request by rhrazdil**
This PR makes the following changes:

- the CNI plugin reads configuration from configMaps.
The plugin gets configuration from configmaps that reference the Network-attachment-definition using the following labels:
  - `cidr-filtering-cni: ""`
  - `<value of the 'name' property in the cni spec>= ""`
Note: Only configmaps that are in the same namespace as the created pod are taken into account.

A configmap may reference multiple Network-attachment-definition specs. Also, single NAD spec may be reference by multiple configmaps.

- the CNI plugin doesn't create empty chains.
- The priority of base chain hooks is set to -1, to avoid issue with packets traversing `inet` table instead of `bridge` table
  See opened BZ for more details: https://bugzilla.redhat.com/show_bug.cgi?id=2099732
- Align allowing all/none ports/subnets with network policy API:
  - When `ports` or `subnets` key is not specified, it translates to not allowing any port/subnet.
  - When `ports` or `subnets` key is empty array, it translates to all ports/subnets allowed

example:
```yaml
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: local-network
  namespace: default
  labels:
    cidr-filtering-cni: ""
    br1-with-cidr-filtering: ""
data:
  config.json: |
    {
      "ingress": {
        "subnets": [
          {
            "subnet": {
              "cidr": "192.168.66.0/24",
              "except": [
                "192.168.66.105"
              ]
            }
          }
        ]
      },
      "egress": {
        "subnets": [
          {
            "subnet": {
              "cidr": "192.168.66.0/24",
              "except": [
                "192.168.66.105"
              ]
            }
          }
        ]
      }
    }
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: network-dhcp
  namespace: default
  labels:
    cidr-filtering-cni: ""
    br1-with-cidr-filtering: ""
data:
  config.json: |
    {
      "egress": {
        "subnets": [
          {
            "subnet": {
              "cidr": "255.255.255.255/32",
              "except": []
            }
          }
        ]
      }
    }
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: google-dns
  namespace: default
  labels:
    cidr-filtering-cni: ""
    br1-with-cidr-filtering: ""
data:
  config.json: |
    {
      "ingress": {
        "subnets": [
          {
            "subnet": {
              "cidr": "8.8.4.4/24",
              "except": [
                "8.8.4.5"
              ]
            }
          }
        ]
      },
      "egress": {
        "subnets": [
          {
            "subnet": {
              "cidr": "8.8.4.4/24"
            }
          }
        ]
      }
    }
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: all-ports
  namespace: default
  labels:
    cidr-filtering-cni: ""
    br1-with-cidr-filtering: ""
data:
  config.json: |
    {
      "ingress": {
        "ports": [
          {
          "protocol": "TCP",
            "port": "1-65535"
          },
          {
          "protocol": "udp",
            "port": "1-65535"
          }
        ]
      },
      "egress": {
        "ports": [
          {
            "protocol": "TCP",
            "port": "1-65535"
          },
          {
            "protocol": "udp",
            "port": "1-65535"
          }
        ]
      }
    }
---
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: a-configmapped-network
spec:
  config: '{
    "cniVersion": "0.3.1",
    "name": "br1-with-cidr-filtering",
    "plugins": [
      {
        "type": "cnv-bridge",
        "bridge": "br1"
      },
      {
        "type": "cidr-filtering-cni"
      }
    ]
  }'
```

Resulting rules
```
table bridge filter {
	chain prerouting {
		type filter hook prerouting priority -1; policy accept;
		ether type arp counter packets 15 bytes 420 accept
		iifname "net1" counter packets 31 bytes 3648 jump m-ingress
	}

	chain postrouting {
		type filter hook postrouting priority -1; policy accept;
		ether type arp counter packets 15 bytes 420 accept
		oifname "net1" counter packets 62 bytes 5650 jump m-egress
	}

	chain m-ingress {
		counter packets 31 bytes 3648 meta mark set meta mark & 0xfffcffff
		iifname "net1" counter packets 31 bytes 3648 jump m-ee129-ingress
		iifname "net1" counter packets 31 bytes 3648 jump m-9b07b-ingress
		iifname "net1" counter packets 31 bytes 3648 jump m-6a1c3-ingress
		meta mark & 0x00030000 == 0x00030000 counter packets 7 bytes 1604 return
		counter packets 24 bytes 2044 drop
	}

	chain m-egress {
		counter packets 62 bytes 5650 meta mark set meta mark & 0xfffcffff
		oifname "net1" counter packets 62 bytes 5650 jump m-ee129-egress
		oifname "net1" counter packets 62 bytes 5650 jump m-9b07b-egress
		oifname "net1" counter packets 62 bytes 5650 jump m-6a1c3-egress
		oifname "net1" counter packets 62 bytes 5650 jump m-61b14-egress
		meta mark & 0x00030000 == 0x00030000 counter packets 8 bytes 1590 return
		counter packets 54 bytes 4060 drop
	}

	chain m-ee129-ingress {
		counter packets 31 bytes 3648 jump m-ee129-ingress-ports
	}

	chain m-ee129-ingress-ports {
		iifname "net1" tcp dport 1-65535 counter packets 0 bytes 0 meta mark set meta mark | 0x00010000
		iifname "net1" udp dport 1-65535 counter packets 7 bytes 1604 meta mark set meta mark | 0x00010000
	}

	chain m-ee129-egress {
		counter packets 62 bytes 5650 jump m-ee129-egress-ports
	}

	chain m-ee129-egress-ports {
		oifname "net1" tcp dport 1-65535 counter packets 0 bytes 0 meta mark set meta mark | 0x00010000
		oifname "net1" udp dport 1-65535 counter packets 44 bytes 4326 meta mark set meta mark | 0x00010000
	}

	chain m-9b07b-ingress {
		counter packets 31 bytes 3648 jump m-9b07b-ingress-subnets
	}

	chain m-9b07b-ingress-subnets {
		iifname "net1" ip saddr 8.8.4.5 counter packets 0 bytes 0 drop
		iifname "net1" ip saddr 8.8.4.0/24 counter packets 0 bytes 0 meta mark set meta mark | 0x00020000
	}

	chain m-9b07b-egress {
		counter packets 62 bytes 5650 jump m-9b07b-egress-subnets
	}

	chain m-9b07b-egress-subnets {
		oifname "net1" ip daddr 8.8.4.0/24 counter packets 0 bytes 0 meta mark set meta mark | 0x00020000
	}

	chain m-6a1c3-ingress {
		counter packets 31 bytes 3648 jump m-6a1c3-ingress-subnets
	}

	chain m-6a1c3-ingress-subnets {
		iifname "net1" ip saddr 192.168.66.105 counter packets 0 bytes 0 drop
		iifname "net1" ip saddr 192.168.66.0/24 counter packets 29 bytes 3416 meta mark set meta mark | 0x00020000
	}

	chain m-6a1c3-egress {
		counter packets 62 bytes 5650 jump m-6a1c3-egress-subnets
	}

	chain m-6a1c3-egress-subnets {
		oifname "net1" ip daddr 192.168.66.105 counter packets 0 bytes 0 drop
		oifname "net1" ip daddr 192.168.66.0/24 counter packets 4 bytes 268 meta mark set meta mark | 0x00020000
	}

	chain m-61b14-egress {
		counter packets 62 bytes 5650 jump m-61b14-egress-subnets
	}

	chain m-61b14-egress-subnets {
		oifname "net1" ip daddr 255.255.255.255 counter packets 4 bytes 1322 meta mark set meta mark | 0x00020000
	}
}

```

### CODE REVIEW

**Comment by maiqueb** [comment]
On your PR description you say:
> The priority of base chain hooks is set to -1, to avoid issue with packets traversing inet table instead of bridge table

Is there an opened issue for this ? Tracking it would be good.

**Comment by maiqueb** [comment]
I have yet to go through the code, but I think you're missing a way to express deny-everything, similar to what [network policies do](https://kubernetes.io/docs/concepts/services-networking/network-policies/#the-two-sorts-of-pod-isolation):
> By default, a pod is non-isolated for egress; all outbound connections are allowed. **A pod is isolated for egress if there is any NetworkPolicy that both selects the pod and has "Egress" in its policyTypes**; we say that such a policy applies to the pod for egress. When a pod is isolated for egress, the only allowed connections from the pod are those allowed by the egress list of some NetworkPolicy that applies to the pod for egress. The effects of those egress lists combine additively.

> By default, a pod is non-isolated for ingress; all inbound connections are allowed. A pod is isolated for ingress if there is any NetworkPolicy that both selects the pod and has "Ingress" in its policyTypes; we say that such a policy applies to the pod for ingress. When a pod is isolated for ingress, the only allowed connections into the pod are those from the pod's node and those allowed by the ingress list of some NetworkPolicy that applies to the pod for ingress. The effects of those ingress lists combine additively.

The idea is provisioning a policy pointing at the pod blocks everything; then you specific punch holes in it, through which traffic flows.

Using your API, I would suggest the following policy to mean deny everything for a pod:
```
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: local-network
  namespace: default
  labels:
    cidr-filtering-cni: ""
    br1-with-cidr-filtering: ""
data:
  config.json: |
    {
      "ingress": { }
    }
```

@rhrazdil

File: cidr-filtering-cni [hunk 1]
```diff
@@ -7,12 +7,12 @@ logFile="${LOGFILE:-/var/log/cidr-filtering-cni.log}"
 
 exec 2>> $logFile
 
-NFT_TABLE_NETDEV=netdev
-NFT_TABLE_BRIDGE=bridge
+NFT_INGRESS_TABLE_TYPE=bridge
+NFT_EGRESS_TABLE_TYPE=bridge
```
Comment (maiqueb) [inline cidr-filtering-cni]: would it make sense to have a single variable for the family ? Both ingress and egress must use the same family ...

File: README.md [hunk 2]
```diff
@@ -10,81 +10,84 @@ This plugin allows user to define allow list for ingress and egress packet filte
 
 ## Usage
 
-By default, all ingress and egress traffic is dropped.
-To allow specific ingress and egress CIDR blocks, create a NetworkAttachmentDefinition. See the following example:
+When using the cidr-filtering-cni plugin, all ingress and egress traffic is dropped by default.
+To allow specific ingress and egress CIDR blocks/ports, create configMaps(s) referencing a NetworkAttachmentDefinition by having a label the NetworkAttachmentDefinition spec name.
+Each configMap additionally needs to have "cidr-filtering-cni" label.
```
Comment (maiqueb) [inline README.md:15]: Please add an example to the PR showing how to drop all ingress and egress traffic.

File: README.md [hunk 3]
```diff
@@ -10,81 +10,84 @@ This plugin allows user to define allow list for ingress and egress packet filte
...
-        }
         ]
       }
     }
-  ]
-}'
 ```
 
 Description of API fields:
+<<<<<<< Updated upstream
```
Comment (maiqueb) [inline README.md]: 0_o

Rebase gone wrong :)

**Review by maiqueb — CHANGES_REQUESTED** [review]
Minor things.

I have some doubts about limiting the configurations to the pod's namespace, but it does make sense.

It does limit the policies re-usability though. 

... let me think it through a bit first.

**Comment by rhrazdil** [comment]
> On your PR description you say:
>>The priority of base chain hooks is set to -1, to avoid issue with packets traversing inet table instead of bridge table

> Is there an opened issue for this ? Tracking it would be good.

Linked [BZ](https://bugzilla.redhat.com/show_bug.cgi?id=2099732) in the PR description.

File: README.md (hunk 3 above)
Comment (rhrazdil) [reply README.md]: Ops, fixed

File: README.md (hunk 2 above)
Comment (rhrazdil) [reply README.md:15]: Done

File: cidr-filtering-cni (hunk 1 above)
Comment (rhrazdil) [reply cidr-filtering-cni]: Yes, Done

**Review by maiqueb — APPROVED** [review]
Thanks.

**Pull request merged**